File: rampardos-render-worker-rs/src/protocol.rs
//! Frame protocol for talking to the Go orchestrator.
//!
//! Wire format (matches `rampardos/internal/services/renderer/protocol.go`
//! and the header comment of `rampardos-render-worker/render-worker.js`):
//!
//!   | 1 byte type | 4 bytes BE uint32 length | length bytes payload |
//!
//! Frame types:
//!   'R' — request  (Go → worker): JSON render request
//!   'K' — ok       (worker → Go): raw RGBA response body
//!   'E' — error    (worker → Go): UTF-8 error message
//!   'H' — handshake (worker → Go): JSON {pid, style} — written once at startup

use std::io::{self, Read, Write};

pub const FRAME_REQUEST: u8 = b'R';
pub const FRAME_OK: u8 = b'K';
pub const FRAME_ERROR: u8 = b'E';
pub const FRAME_HANDSHAKE: u8 = b'H';

/// Cap frame payloads at 64 MiB — generous for 4096×4096 RGBA. Matches the
/// Go orchestrator's `MaxFrameSize` constant.
pub const MAX_FRAME_SIZE: u32 = 64 * 1024 * 1024;

pub fn write_frame<W: Write>(w: &mut W, typ: u8, payload: &[u8]) -> io::Result<()> {
    if payload.len() > MAX_FRAME_SIZE as usize {
        return Err(io::Error::new(
            io::ErrorKind::InvalidInput,
            format!("frame payload too large: {} > {}", payload.len(), MAX_FRAME_SIZE),
        ));
    }
    let mut header = [0u8; 5];
    header[0] = typ;
    header[1..5].copy_from_slice(&(payload.len() as u32).to_be_bytes());
    w.write_all(&header)?;
    if !payload.is_empty() {
        w.write_all(payload)?;
    }
    w.flush()
}

pub enum ReadFrameResult {
    Frame { typ: u8, payload: Vec<u8> },
    Eof,
}

pub fn read_frame<R: Read>(r: &mut R) -> io::Result<ReadFrameResult> {
    let mut header = [0u8; 5];
    match r.read_exact(&mut header) {
        Ok(()) => {}
        Err(e) if e.kind() == io::ErrorKind::UnexpectedEof => {
            return Ok(ReadFrameResult::Eof);
        }
        Err(e) => return Err(e),
    }
    let typ = header[0];
    let len = u32::from_be_bytes([header[1], header[2], header[3], header[4]]);
    if len > MAX_FRAME_SIZE {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            format!("frame payload too large: {} > {}", len, MAX_FRAME_SIZE),
        ));
    }
    let mut payload = vec![0u8; len as usize];
    if len > 0 {
        r.read_exact(&mut payload)?;
    }
    Ok(ReadFrameResult::Frame { typ, payload })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn roundtrip_frame() {
        let mut buf: Vec<u8> = Vec::new();
        write_frame(&mut buf, FRAME_OK, b"hello").unwrap();

        let mut cursor = std::io::Cursor::new(&buf);
        let frame = read_frame(&mut cursor).unwrap();
        match frame {
            ReadFrameResult::Frame { typ, payload } => {
                assert_eq!(typ, FRAME_OK);
                assert_eq!(payload, b"hello");
            }
            ReadFrameResult::Eof => panic!("unexpected EOF"),
        }
    }

    #[test]
    fn empty_payload() {
        let mut buf: Vec<u8> = Vec::new();
        write_frame(&mut buf, FRAME_HANDSHAKE, b"").unwrap();
        assert_eq!(buf.len(), 5);
        assert_eq!(buf[0], FRAME_HANDSHAKE);
        assert_eq!(&buf[1..5], &[0, 0, 0, 0]);
    }

    #[test]
    fn read_eof_on_empty_stream() {
        let mut cursor = std::io::Cursor::new(&[][..]);
        assert!(matches!(read_frame(&mut cursor).unwrap(), ReadFrameResult::Eof));
    }

    #[test]
    fn length_is_big_endian() {
        // Length 256 should encode as 00 00 01 00.
        let mut buf: Vec<u8> = Vec::new();
        write_frame(&mut buf, FRAME_OK, &vec![0; 256]).unwrap();
        assert_eq!(&buf[1..5], &[0, 0, 1, 0]);
    }
}

File: rampardos/internal/services/renderer/protocol.go
package renderer

import (
	"encoding/binary"
	"fmt"
	"io"
)

// Frame types. All frames are length-prefixed: one byte type, four
// bytes big-endian uint32 length, then the payload. The worker and
// orchestrator must agree on these constants; they are also documented
// at the top of rampardos-render-worker/render-worker.js.
const (
	FrameRequest   byte = 'R' // Go -> worker: JSON render request
	FrameOK        byte = 'K' // worker -> Go: raw RGBA response
	FrameError     byte = 'E' // worker -> Go: UTF-8 error message
	FrameHandshake byte = 'H' // worker -> Go: initial readiness handshake
)

// FrameHeaderSize is the number of bytes in a frame header (1 type + 4 length).
const FrameHeaderSize = 5

// MaxFrameSize caps payload length to prevent a misbehaving worker
// from exhausting memory by sending huge frames.
const MaxFrameSize = 64 * 1024 * 1024 // 64 MiB — generous for 4096×4096 RGBA (64 MiB)

// WriteFrame writes a single framed message to w.
func WriteFrame(w io.Writer, typ byte, payload []byte) error {
	if len(payload) > MaxFrameSize {
		return fmt.Errorf("renderer: frame payload too large: %d > %d", len(payload), MaxFrameSize)
	}
	header := [FrameHeaderSize]byte{}
	header[0] = typ
	binary.BigEndian.PutUint32(header[1:], uint32(len(payload)))
	if _, err := w.Write(header[:]); err != nil {
		return err
	}
	if len(payload) > 0 {
		if _, err := w.Write(payload); err != nil {
			return err
		}
	}
	return nil
}

// ReadFrame reads a single framed message from r.
// Returns io.EOF only when no bytes at all are available for a new
// frame (clean shutdown); a partial header or payload returns a
// specific error.
func ReadFrame(r io.Reader) (typ byte, payload []byte, err error) {
	var header [FrameHeaderSize]byte
	n, err := io.ReadFull(r, header[:])
	if n == 0 && err == io.EOF {
		return 0, nil, io.EOF
	}
	if err != nil {
		return 0, nil, fmt.Errorf("renderer: read frame header: %w", err)
	}

	typ = header[0]
	length := binary.BigEndian.Uint32(header[1:])
	if length > MaxFrameSize {
		return 0, nil, fmt.Errorf("renderer: frame payload too large: %d > %d", length, MaxFrameSize)
	}

	if length == 0 {
		return typ, []byte{}, nil
	}

	payload = make([]byte, length)
	if _, err := io.ReadFull(r, payload); err != nil {
		return 0, nil, fmt.Errorf("renderer: read frame payload: %w", err)
	}

	return typ, payload, nil
}
